
 # Dealing With Null Values Solution Notebook.
 <br>
 <br>
 It's time to get your hands dirty dude!!
 -Bit mL

# Import Libraries

In [13]:
import numpy as np
import pandas as pd

 Python list <b>headers</b> containing name of headers 


In [10]:
# These are the headers which you need to link to your dataset.
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]

# Import Datasets

Use the Pandas method <b>read_csv()</b> to load the data from the web address. Set the parameter  "names" equal to the Python list "headers".


In [15]:
df = pd.read_csv('car_dataset.data', names = headers)
df

symboling normalized-losses         make fuel-type aspiration  \
0            3                 ?  alfa-romero       gas        std   
1            3                 ?  alfa-romero       gas        std   
2            1                 ?  alfa-romero       gas        std   
3            2               164         audi       gas        std   
4            2               164         audi       gas        std   
..         ...               ...          ...       ...        ...   
200         -1                95        volvo       gas        std   
201         -1                95        volvo       gas      turbo   
202         -1                95        volvo       gas        std   
203         -1                95        volvo    diesel      turbo   
204         -1                95        volvo       gas      turbo   

    num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0            two  convertible          rwd           front        88.6  ...   
1            two  convertible          rwd           front        88.6  ...   
2            two    hatchback          rwd           front        94.5  ...   
3           four        sedan          fwd           front        99.8  ...   
4           four        sedan          4wd           front        99.4  ...   
..           ...          ...          ...             ...         ...  ...   
200         four        sedan          rwd           front       109.1  ...   
201         four        sedan          rwd           front       109.1  ...   
202         four        sedan          rwd           front       109.1  ...   
203         four        sedan          rwd           front       109.1  ...   
204         four        sedan          rwd           front       109.1  ...   

     engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0            130         mpfi  3.47    2.68               9.0        111   
1            130         mpfi  3.47    2.68               9.0        111   
2            152         mpfi  2.68    3.47               9.0        154   
3            109         mpfi  3.19    3.40              10.0        102   
4            136         mpfi  3.19    3.40               8.0        115   
..           ...          ...   ...     ...               ...        ...   
200          141         mpfi  3.78    3.15               9.5        114   
201          141         mpfi  3.78    3.15               8.7        160   
202          173         mpfi  3.58    2.87               8.8        134   
203          145          idi  3.01    3.40              23.0        106   
204          141         mpfi  3.78    3.15               9.5        114   

     peak-rpm city-mpg highway-mpg  price  
0        5000       21          27  13495  
1        5000       21          27  16500  
2        5000       19          26  16500  
3        5500       24          30  13950  
4        5500       18          22  17450  
..        ...      ...         ...    ...  
200      5400       23          28  16845  
201      5300       19          25  19045  
202      5500       18          23  21485  
203      4800       26          27  22470  
204      5400       19          25  22625  

[205 rows x 26 columns]

 Use the method <b>head()</b> to display the first five rows of the dataframe. 


In [14]:
# To see what the data set looks like, we'll use the head() method.
df.head(15)


symboling normalized-losses         make fuel-type aspiration  \
0           3                 ?  alfa-romero       gas        std   
1           3                 ?  alfa-romero       gas        std   
2           1                 ?  alfa-romero       gas        std   
3           2               164         audi       gas        std   
4           2               164         audi       gas        std   
5           2                 ?         audi       gas        std   
6           1               158         audi       gas        std   
7           1                 ?         audi       gas        std   
8           1               158         audi       gas      turbo   
9           0                 ?         audi       gas      turbo   
10          2               192          bmw       gas        std   
11          0               192          bmw       gas        std   
12          0               188          bmw       gas        std   
13          0               188          bmw       gas        std   
14          1                 ?          bmw       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two  convertible          rwd           front        88.6  ...   
2           two    hatchback          rwd           front        94.5  ...   
3          four        sedan          fwd           front        99.8  ...   
4          four        sedan          4wd           front        99.4  ...   
5           two        sedan          fwd           front        99.8  ...   
6          four        sedan          fwd           front       105.8  ...   
7          four        wagon          fwd           front       105.8  ...   
8          four        sedan          fwd           front       105.8  ...   
9           two    hatchback          4wd           front        99.5  ...   
10          two        sedan          rwd           front       101.2  ...   
11         four        sedan          rwd           front       101.2  ...   
12          two        sedan          rwd           front       101.2  ...   
13         four        sedan          rwd           front       101.2  ...   
14         four        sedan          rwd           front       103.5  ...   

    engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0           130         mpfi  3.47    2.68               9.0        111   
1           130         mpfi  3.47    2.68               9.0        111   
2           152         mpfi  2.68    3.47               9.0        154   
3           109         mpfi  3.19    3.40              10.0        102   
4           136         mpfi  3.19    3.40               8.0        115   
5           136         mpfi  3.19    3.40               8.5        110   
6           136         mpfi  3.19    3.40               8.5        110   
7           136         mpfi  3.19    3.40               8.5        110   
8           131         mpfi  3.13    3.40               8.3        140   
9           131         mpfi  3.13    3.40               7.0        160   
10          108         mpfi  3.50    2.80               8.8        101   
11          108         mpfi  3.50    2.80               8.8        101   
12          164         mpfi  3.31    3.19               9.0        121   
13          164         mpfi  3.31    3.19               9.0        121   
14          164         mpfi  3.31    3.19               9.0        121   

    peak-rpm city-mpg highway-mpg  price  
0       5000       21          27  13495  
1       5000       21          27  16500  
2       5000       19          26  16500  
3       5500       24          30  13950  
4       5500       18          22  17450  
5       5500       19          25  15250  
6       5500       19          25  17710  
7       5500       19          25  18920  
8       5500       17          20  23875  
9       5500       

As we can see, several question marks appeared in the dataframe; those are missing values which may hinder our further analysis. 

<div>So, how do we identify all those missing values and deal with them?</div> 

<b>How to work with missing data?</b>

Steps for working with missing data:

<ol>
    <li>Identify missing data</li>
    <li>deal with missing data</li>
    <li>correct data format</li>
</ol>


<h1 id="identify_handle_missing_values">Identify and handle missing values</h1>

<h2 id="identify_missing_values">Identify missing values</h2>
<h3>Convert "?" to NaN</h3>
In the car dataset, missing data comes with the question mark "?".
We replace "?" with NaN (Not a Number), which is Python's default missing value marker, for reasons of computational speed and convenience. Here we use the function: 
 <pre>.replace(A, B, inplace = True) </pre>
to replace A by B


In [20]:
# replace "?" to NaN
df.replace("?", np.nan, inplace=True)
df

symboling normalized-losses         make fuel-type aspiration  \
0            3               NaN  alfa-romero       gas        std   
1            3               NaN  alfa-romero       gas        std   
2            1               NaN  alfa-romero       gas        std   
3            2               164         audi       gas        std   
4            2               164         audi       gas        std   
..         ...               ...          ...       ...        ...   
200         -1                95        volvo       gas        std   
201         -1                95        volvo       gas      turbo   
202         -1                95        volvo       gas        std   
203         -1                95        volvo    diesel      turbo   
204         -1                95        volvo       gas      turbo   

    num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0            two  convertible          rwd           front        88.6  ...   
1            two  convertible          rwd           front        88.6  ...   
2            two    hatchback          rwd           front        94.5  ...   
3           four        sedan          fwd           front        99.8  ...   
4           four        sedan          4wd           front        99.4  ...   
..           ...          ...          ...             ...         ...  ...   
200         four        sedan          rwd           front       109.1  ...   
201         four        sedan          rwd           front       109.1  ...   
202         four        sedan          rwd           front       109.1  ...   
203         four        sedan          rwd           front       109.1  ...   
204         four        sedan          rwd           front       109.1  ...   

     engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0            130         mpfi  3.47    2.68               9.0        111   
1            130         mpfi  3.47    2.68               9.0        111   
2            152         mpfi  2.68    3.47               9.0        154   
3            109         mpfi  3.19    3.40              10.0        102   
4            136         mpfi  3.19    3.40               8.0        115   
..           ...          ...   ...     ...               ...        ...   
200          141         mpfi  3.78    3.15               9.5        114   
201          141         mpfi  3.78    3.15               8.7        160   
202          173         mpfi  3.58    2.87               8.8        134   
203          145          idi  3.01    3.40              23.0        106   
204          141         mpfi  3.78    3.15               9.5        114   

     peak-rpm city-mpg highway-mpg  price  
0        5000       21          27  13495  
1        5000       21          27  16500  
2        5000       19          26  16500  
3        5500       24          30  13950  
4        5500       18          22  17450  
..        ...      ...         ...    ...  
200      5400       23          28  16845  
201      5300       19          25  19045  
202      5500       18          23  21485  
203      4800       26          27  22470  
204      5400       19          25  22625  

[205 rows x 26 columns]

Identify_missing_values

<h2>Evaluating for Missing Data</h2>

The missing values are converted to default. We use the following functions to identify these missing values. There are two methods to detect missing data:

<ol>
    <li><b>.isnull()</b></li>
    <li><b>.notnull()</b></li>
</ol>
The output is a boolean value indicating whether the value that is passed into the argument is in fact missing data.


"True" stands for missing value, while "False" stands for not missing value.


In [21]:
df.isnull()

symboling  normalized-losses   make  fuel-type  aspiration  num-of-doors  \
0        False               True  False      False       False         False   
1        False               True  False      False       False         False   
2        False               True  False      False       False         False   
3        False              False  False      False       False         False   
4        False              False  False      False       False         False   
..         ...                ...    ...        ...         ...           ...   
200      False              False  False      False       False         False   
201      False              False  False      False       False         False   
202      False              False  False      False       False         False   
203      False              False  False      False       False         False   
204      False              False  False      False       False         False   

     body-style  drive-wheels  engine-location  wheel-base  ...  engine-size  \
0         False         False            False       False  ...        False   
1         False         False            False       False  ...        False   
2         False         False            False       False  ...        False   
3         False         False            False       False  ...        False   
4         False         False            False       False  ...        False   
..          ...           ...              ...         ...  ...          ...   
200       False         False            False       False  ...        False   
201       False         False            False       False  ...        False   
202       False         False            False       False  ...        False   
203       False         False            False       False  ...        False   
204       False         False            False       False  ...        False   

     fuel-system   bore  stroke  compression-ratio  horsepower  peak-rpm  \
0          False  False   False              False       False     False   
1          False  False   False              False       False     False   
2          False  False   False              False       False     False   
3          False  False   False              False       False     False   
4          False  False   False              False       False     False   
..           ...    ...     ...                ...         ...       ...   
200        False  False   False              False       False     False   
201        False  False   False              False       False     False   
202        False  False   False              False       False     False   
203        False  False   False              False       False     False   
204        False  False   False              False       False     False   

     city-mpg  highway-mpg  price  
0       False        False  False  
1       False        False  False  
2       False        False  False  
3       False        False  False  
4       False        False  False  
..        ...          ...    ...  
200     False        False  False  
201     False        False  False  
202     False        False  False  
203     False        False  False  
204     False        False  False  

[205 rows x 26 columns]

<h2>Count missing values in each column</h2>
<p>
Using a for loop in Python, we can quickly figure out the number of missing values in each column. As mentioned above, "True" represents a missing value, "False"  means the value is present in the dataset.  In the body of the for loop the method  ".value_counts()"  counts the number of "True" values. 
</p>


In [22]:
df.notnull()

symboling  normalized-losses  make  fuel-type  aspiration  num-of-doors  \
0         True              False  True       True        True          True   
1         True              False  True       True        True          True   
2         True              False  True       True        True          True   
3         True               True  True       True        True          True   
4         True               True  True       True        True          True   
..         ...                ...   ...        ...         ...           ...   
200       True               True  True       True        True          True   
201       True               True  True       True        True          True   
202       True               True  True       True        True          True   
203       True               True  True       True        True          True   
204       True               True  True       True        True          True   

     body-style  drive-wheels  engine-location  wheel-base  ...  engine-size  \
0          True          True             True        True  ...         True   
1          True          True             True        True  ...         True   
2          True          True             True        True  ...         True   
3          True          True             True        True  ...         True   
4          True          True             True        True  ...         True   
..          ...           ...              ...         ...  ...          ...   
200        True          True             True        True  ...         True   
201        True          True             True        True  ...         True   
202        True          True             True        True  ...         True   
203        True          True             True        True  ...         True   
204        True          True             True        True  ...         True   

     fuel-system  bore  stroke  compression-ratio  horsepower  peak-rpm  \
0           True  True    True               True        True      True   
1           True  True    True               True        True      True   
2           True  True    True               True        True      True   
3           True  True    True               True        True      True   
4           True  True    True               True        True      True   
..           ...   ...     ...                ...         ...       ...   
200         True  True    True               True        True      True   
201         True  True    True               True        True      True   
202         True  True    True               True        True      True   
203         True  True    True               True        True      True   
204         True  True    True               True        True      True   

     city-mpg  highway-mpg  price  
0        True         True   True  
1        True         True   True  
2        True         True   True  
3        True         True   True  
4        True         True   True  
..        ...          ...    ...  
200      True         True   True  
201      True         True   True  
202      True         True   True  
203      True         True   True  
204      True         True   True  

[205 rows x 26 columns]

In [23]:
missing_data = df.isnull()
missing_data

symboling  normalized-losses   make  fuel-type  aspiration  num-of-doors  \
0        False               True  False      False       False         False   
1        False               True  False      False       False         False   
2        False               True  False      False       False         False   
3        False              False  False      False       False         False   
4        False              False  False      False       False         False   
..         ...                ...    ...        ...         ...           ...   
200      False              False  False      False       False         False   
201      False              False  False      False       False         False   
202      False              False  False      False       False         False   
203      False              False  False      False       False         False   
204      False              False  False      False       False         False   

     body-style  drive-wheels  engine-location  wheel-base  ...  engine-size  \
0         False         False            False       False  ...        False   
1         False         False            False       False  ...        False   
2         False         False            False       False  ...        False   
3         False         False            False       False  ...        False   
4         False         False            False       False  ...        False   
..          ...           ...              ...         ...  ...          ...   
200       False         False            False       False  ...        False   
201       False         False            False       False  ...        False   
202       False         False            False       False  ...        False   
203       False         False            False       False  ...        False   
204       False         False            False       False  ...        False   

     fuel-system   bore  stroke  compression-ratio  horsepower  peak-rpm  \
0          False  False   False              False       False     False   
1          False  False   False              False       False     False   
2          False  False   False              False       False     False   
3          False  False   False              False       False     False   
4          False  False   False              False       False     False   
..           ...    ...     ...                ...         ...       ...   
200        False  False   False              False       False     False   
201        False  False   False              False       False     False   
202        False  False   False              False       False     False   
203        False  False   False              False       False     False   
204        False  False   False              False       False     False   

     city-mpg  highway-mpg  price  
0       False        False  False  
1       False        False  False  
2       False        False  False  
3       False        False  False  
4       False        False  False  
..        ...          ...    ...  
200     False        False  False  
201     False        False  False  
202     False        False  False  
203     False        False  False  
204     False        False  False  

[205 rows x 26 columns]

In [26]:
missing_data = df.isnull()
for column in headers:
    print(column)
    print(missing_data[column].value_counts())
    print(" ")

symboling
symboling
False    205
Name: count, dtype: int64
 
normalized-losses
normalized-losses
False    164
True      41
Name: count, dtype: int64
 
make
make
False    205
Name: count, dtype: int64
 
fuel-type
fuel-type
False    205
Name: count, dtype: int64
 
aspiration
aspiration
False    205
Name: count, dtype: int64
 
num-of-doors
num-of-doors
False    203
True       2
Name: count, dtype: int64
 
body-style
body-style
False    205
Name: count, dtype: int64
 
drive-wheels
drive-wheels
False    205
Name: count, dtype: int64
 
engine-location
engine-location
False    205
Name: count, dtype: int64
 
wheel-base
wheel-base
False    205
Name: count, dtype: int64
 
length
length
False    205
Name: count, dtype: int64
 
width
width
False    205
Name: count, dtype: int64
 
height
height
False    205
Name: count, dtype: int64
 
curb-weight
curb-weight
False    205
Name: count, dtype: int64
 
engine-type
engine-type
False    205
Name: count, dtype: int64
 
num-of-cylinders
num-of-cylinders
F

Based on the summary above, each column has 205 rows of data, seven columns containing missing data:

<ol>
    <li>"normalized-losses": 41 missing data</li>
    <li>"num-of-doors": 2 missing data</li>
    <li>"bore": 4 missing data</li>
    <li>"stroke" : 4 missing data</li>
    <li>"horsepower": 2 missing data</li>
    <li>"peak-rpm": 2 missing data</li>
    <li>"price": 4 missing data</li>
</ol>


<h2 id="deal_missing_values">Deal with missing data</h2>
<br>
<h3>How to deal with missing data?</h3>

<ol>
    <li>drop data<br>
        a. drop the whole row<br>
        b. drop the whole column
    </li>
    <li>replace data<br>
        a. replace it by mean<br>
        b. replace it by frequency<br>
    </li>
</ol>


Whole columns should be dropped only if most entries in the column are empty. In our dataset, none of the columns are empty enough to drop entirely.
We have some freedom in choosing which method to replace data; however, some methods may seem more reasonable than others. We will apply each method to many different columns:

<b>Replace by mean:</b>

<ul>
    <li>"normalized-losses": 41 missing data, replace them with mean</li>
    <li>"stroke": 4 missing data, replace them with mean</li>
    <li>"bore": 4 missing data, replace them with mean</li>
    <li>"horsepower": 2 missing data, replace them with mean</li>
    <li>"peak-rpm": 2 missing data, replace them with mean</li>
</ul>

<b>Replace by frequency:</b>

<ul>
    <li>"num-of-doors": 2 missing data, replace them with "four". 
        <ul>
            <li>Reason: 84% sedans is four doors. Since four doors is most frequent, it is most likely to occur</li>
        </ul>
    </li>
</ul>

<b>Drop the whole row:</b>

<ul>
    <li>"price": 4 missing data, simply delete the whole row
        <ul>
            <li>Reason: price is what we want to predict. Any data entry without price data cannot be used for prediction; therefore any row now without price data is not useful to us</li>
        </ul>
    </li>
</ul>


<h3>Calculate the average of the normalized-losses column </h3>


In [29]:
avg_num_loss=df['normalized-losses']. astype("float").mean()
avg_num_loss

122.0

<h3>Replace "NaN" by mean value in "normalized-losses" column</h3>


In [71]:
df['normalized-losses'].replace(np.nan, avg_num_loss, inplace=True)
df.head(10)

symboling normalized-losses         make fuel-type aspiration  \
0           3             122.0  alfa-romero       gas        std   
1           3             122.0  alfa-romero       gas        std   
2           1             122.0  alfa-romero       gas        std   
3           2               164         audi       gas        std   
4           2               164         audi       gas        std   
5           2             122.0         audi       gas        std   
6           1               158         audi       gas        std   
7           1             122.0         audi       gas        std   
8           1               158         audi       gas      turbo   
10          2               192          bmw       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two  convertible          rwd           front        88.6  ...   
2           two    hatchback          rwd           front        94.5  ...   
3          four        sedan          fwd           front        99.8  ...   
4          four        sedan          4wd           front        99.4  ...   
5           two        sedan          fwd           front        99.8  ...   
6          four        sedan          fwd           front       105.8  ...   
7          four        wagon          fwd           front       105.8  ...   
8          four        sedan          fwd           front       105.8  ...   
10          two        sedan          rwd           front       101.2  ...   

    engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0           130         mpfi  3.47    2.68               9.0        111   
1           130         mpfi  3.47    2.68               9.0        111   
2           152         mpfi  2.68    3.47               9.0        154   
3           109         mpfi  3.19    3.40              10.0        102   
4           136         mpfi  3.19    3.40               8.0        115   
5           136         mpfi  3.19    3.40               8.5        110   
6           136         mpfi  3.19    3.40               8.5        110   
7           136         mpfi  3.19    3.40               8.5        110   
8           131         mpfi  3.13    3.40               8.3        140   
10          108         mpfi  3.50    2.80               8.8        101   

    peak-rpm city-mpg highway-mpg  price  
0       5000       21          27  13495  
1       5000       21          27  16500  
2       5000       19          26  16500  
3       5500       24          30  13950  
4       5500       18          22  17450  
5       5500       19          25  15250  
6       5500       19          25  17710  
7       5500       19          25  18920  
8       5500       17          20  23875  
10      5800       23          29  16430  

[10 rows x 26 columns]

<h3>Calculate the mean value for 'bore' column</h3>


In [37]:
avg_bore=df['stroke'].astype('float').mean(axis=0)
print("Average of stroke:" , avg_bore)

Average of stroke: 3.2554228855721394


<h3>Replace NaN by mean value for bore column</h3>


<h3>Calculate the mean value for 'stroke' column</h3>

In [42]:
# Write your code below and press Shift+Enter to execute 
avg_bore=df['stroke'].astype('float').mean(axis=0)
print("Average of stroke:" , avg_bore)

Average of stroke: 3.2554228855721394


<h3>Replace NaN by mean value for stroke column</h3>

In [70]:
import warnings
warnings.filterwarnings("ignore")
df['stroke'].replace(np.nan, avg_bore, inplace=True)
df.head(40)


symboling normalized-losses         make fuel-type aspiration  \
0           3             122.0  alfa-romero       gas        std   
1           3             122.0  alfa-romero       gas        std   
2           1             122.0  alfa-romero       gas        std   
3           2               164         audi       gas        std   
4           2               164         audi       gas        std   
5           2             122.0         audi       gas        std   
6           1               158         audi       gas        std   
7           1             122.0         audi       gas        std   
8           1               158         audi       gas      turbo   
10          2               192          bmw       gas        std   
11          0               192          bmw       gas        std   
12          0               188          bmw       gas        std   
13          0               188          bmw       gas        std   
14          1             122.0          bmw       gas        std   
15          0             122.0          bmw       gas        std   
16          0             122.0          bmw       gas        std   
17          0             122.0          bmw       gas        std   
18          2               121    chevrolet       gas        std   
19          1                98    chevrolet       gas        std   
20          0                81    chevrolet       gas        std   
21          1               118        dodge       gas        std   
22          1               118        dodge       gas        std   
23          1               118        dodge       gas      turbo   
24          1               148        dodge       gas        std   
25          1               148        dodge       gas        std   
26          1               148        dodge       gas        std   
27          1               148        dodge       gas      turbo   
28         -1               110        dodge       gas        std   
29          3               145        dodge       gas      turbo   
30          2               137        honda       gas        std   
31          2               137        honda       gas        std   
32          1               101        honda       gas        std   
33          1               101        honda       gas        std   
34          1               101        honda       gas        std   
35          0               110        honda       gas        std   
36          0                78        honda       gas        std   
37          0               106        honda       gas        std   
38          0               106        honda       gas        std   
39          0                85        honda       gas        std   
40          0                85        honda       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two  convertible          rwd           front        88.6  ...   
2           two    hatchback          rwd           front        94.5  ...   
3          four        sedan          fwd           front        99.8  ...   
4          four        sedan          4wd           front        99.4  ...   
5           two        sedan          fwd           front        99.8  ...   
6          four        sedan          fwd           front       105.8  ...   
7          four        wagon          fwd           front       105.8  ...   
8          four        sedan          fwd           front       105.8  ...   
10          two        sedan          rwd           front       101.2  ...   
11         four        sedan          rwd           front       101.2  ...   
12          two        sedan          rwd           front       101.2  ...   
13         four        sedan          rwd           front       101.2  ...   
14         four        sedan          rwd           front       103.5  ...   
15  

<details><summary>Click here for the solution</summary>

```python
#Calculate the mean vaule for "stroke" column
avg_stroke = df["stroke"].astype("float").mean(axis = 0)
print("Average of stroke:", avg_stroke)

# replace NaN by mean value in "stroke" column
df["stroke"].replace(np.nan, avg_stroke, inplace = True)
```

</details>


<h4>Calculate the mean value for the  'horsepower' column:</h4>


In [44]:
avg_horsepower=df['horsepower'].astype('float').mean(axis=0)
print("Average of horsepower:" , avg_horsepower)

Average of horsepower: 104.25615763546799


<h4>Replace "NaN" by mean value:</h4>


In [50]:

import warnings
warnings.filterwarnings("ignore")
df['horsepower'].replace(np.nan, avg_horsepower, inplace =True)

0      111
1      111
2      154
3      102
4      115
      ... 
200    114
201    160
202    134
203    106
204    114
Name: horsepower, Length: 205, dtype: object

<h4>Calculate the mean value for 'peak-rpm' column:</h4>


In [51]:
avg_peakrpm=df['peak-rpm'].astype('float').mean(axis=0)
print("Average of peak rpm:" , avg_peakrpm)

Average of peak rpm: 5125.369458128079


<h4>Replace NaN by mean value:</h4>


In [66]:


import warnings
warnings.filterwarnings("ignore")
df['peak-rpm'].replace(np.nan, avg_peakrpm, inplace=True)


<br>

## Alright you have done great job till now, but its not finished yet dude!!!
<br>

<ul>
<li>Now we have done for replacing with mean.</li>
<li>Its time to replace null values with most frequent values </li>
<li>We will do this for num-of-doors column</li>
</ul>

<h4>Find the most frequent value for num of doors</h4>

In [57]:
df['num-of-doors'].value_counts()
df['body-style'].value_counts()

body-style
sedan          96
hatchback      70
wagon          25
hardtop         8
convertible     6
Name: count, dtype: int64

<h4>We can see that four doors are the most common type. We can also use the ".idxmax()" method to calculate for us the most common type automatically:</h4>


In [58]:
df['num-of-doors'].value_counts().idxmax()

'four'

<h4>The replacement procedure is very similar to what we have seen previously</h4>


In [65]:
#replace the missing 'num-of-doors' values by the most frequent 

import warnings
warnings.filterwarnings("ignore")
df['num-of-doors'].replace(np.nan, 'four', inplace =True)

## Finally, let's drop all rows that do not have price data:


In [63]:
# simply drop whole row with NaN in "price" column

df.dropna(subset =['price'], inplace = True, axis =0)
# reset index, because we droped two rows


<h2> Lets Check It</h2>

In [74]:
df.reset_index(drop = True, inplace = True)


In [75]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 201 entries, 0 to 200
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          201 non-null    int64  
 1   normalized-losses  201 non-null    object 
 2   make               201 non-null    object 
 3   fuel-type          201 non-null    object 
 4   aspiration         201 non-null    object 
 5   num-of-doors       201 non-null    object 
 6   body-style         201 non-null    object 
 7   drive-wheels       201 non-null    object 
 8   engine-location    201 non-null    object 
 9   wheel-base         201 non-null    float64
 10  length             201 non-null    float64
 11  width              201 non-null    float64
 12  height             201 non-null    float64
 13  curb-weight        201 non-null    int64  
 14  engine-type        201 non-null    object 
 15  num-of-cylinders   201 non-null    object 
 16  engine-size        201 non

Hurrey!!!
Now we have our dataset with no missing values.
<br>
Isn't It?

<br>
<br>

# REALLY IT WAS A GREAT JOB DUDE!!! 
<h4>LETS START LEARNING DATA FORMATING & DATA STANDARDIZATION.</h4>
<h4>There you will learn how to correct the data types of each columns in proper way. </h4>
<br>
<br>

# Bit ML
<h4>We make complex simple.</h4>
<h4>It's Easy If you learn like this.</h4>

# Thank you So much.